server/main.lua

```diff
@@ -55,47 +40,74 @@ RegisterNetEvent('qb-vehiclekeys:server:setVehLockState', function(vehNetId, sta
     SetVehicleDoorsLocked(NetworkGetEntityFromNetworkId(vehNetId), state)
 end)
 
-lib.callback.register('qbx-vehiclekeys:server:getVehicleKeys', function(source)
-    local citizenid = exports.qbx_core:GetPlayer(source).PlayerData.citizenid
-    local keysList = {}
-    for plate, citizenids in pairs (vehicleList) do
-        if citizenids[citizenid] then
-            keysList[plate] = true
-        end
-    end
-    return keysList
+local function getCitizenId(source)
+    local player = exports.qbx_core:GetPlayer(source)
+
+    if not player then return end
+    local citizenid = player.PlayerData.citizenid
+
+    return citizenid
+end
+
+RegisterNetEvent('qbx-vehiclekeys:server:setPlayerKeys', function()
+    local src = source
+    local citizenid = getCitizenId(src)
+    Player(src).state.keysList = keysList[citizenid]
 end)
 
 -----------------------
 ----   Functions   ----
 -----------------------
-function GiveKeys(id, plate)
-    local citizenid = exports.qbx_core:GetPlayer(id).PlayerData.citizenid
 
-    if not vehicleList[plate] then vehicleList[plate] = {} end
-    vehicleList[plate][citizenid] = true
+-- local function isVehicleVisibleByServer(plate)
+--     local vehicles = GetAllVehicles()
+--     if type(vehicles) == "table" then
+--         for _, vehicle in ipairs(vehicles) do
+--             local p = GetVehicleNumberPlateText(vehicle)
+--             if p == plate then
+--                 Wait(500)
+--                 if GetEntityPopulationType(vehicle) == 7 then
+--                     return true
+--                 end
+--             end
+--         end
+--     end
+-- end
+
+function GiveKeys(source, plate)
+    local keys = Player(source).state.keysList or {}
+
+    if keys[plate] then return end
+    keys[plate] = true
+    Player(source).state:set('keysList', keys, true)
+    local citizenid = getCitizenId(source)
+
+    if not citizenid then return end
+
+    -- if isVehicleVisibleByServer(plate) then
+    if not keysList[citizenid] then
+        keysList[citizenid] = {}
+    end
+
+    keysList[citizenid][plate] = true
```

Comment:
Do we even need this table now that the keys are stored on the player's state bag? This question boils down to do we need to access keys by citizenid, or would serverId suffice.

Reply:
It's quite problematic. if the player disconnects and reconnects, the statebag disappears. Maybe new item "vehicle keys" in inventory?

Reply:
Ah. Maybe we should limit its use to saving/loading on player disconnect/reconnect then? Or maybe there is a broader solution that could be put in core of saving/loading statebags? No action required.

Reply:
Another point to consider is whether this data should be persisted between server restarts?

Reply:
What if in hotwiring replace TriggerServerEvent('qb-vehiclekeys:server:AcquireVehicleKeys') with a statebag change?
Then only the keys assigned by the scripts would be in the server's memory.